packages/application-shell/src/components/custom-view-shell/custom-view-shell.tsx

```diff
@@ -0,0 +1,132 @@
+import {
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+  type ReactNode,
+} from 'react';
+import type { ApplicationWindow } from '@commercetools-frontend/constants';
+import type { TAsyncLocaleDataProps } from '@commercetools-frontend/i18n';
+import LoadingSpinner from '@commercetools-uikit/loading-spinner';
+import ApplicationShellProvider from '../application-shell-provider';
+import CustomViewAuthenticatedShell from '../custom-view-authenticated-shell';
+import { CustomViewContextProvider, TCustomView } from '../custom-view-context';
+import { CUSTOM_VIEWS_EVENTS_NAMES } from './constants';
+
+declare let window: ApplicationWindow;
+
+type THostContext = {
+  hostUrl: string;
+  userLocale: string;
+  dataLocale: string;
+  customViewConfig: TCustomView;
+  projectKey?: string;
+};
+
+type THostEventData = {
+  source: string;
+  destination: string;
+  eventName: string;
+  eventData: {
+    context: THostContext;
+  };
+};
+
+type TCustomViewShellProps = {
+  customViewId: string;
+  messages: TAsyncLocaleDataProps['applicationMessages'];
+  children: ReactNode;
+};
+
+function CustomViewShell(props: TCustomViewShellProps) {
+  const [hostContext, setHostContext] = useState<THostContext>();
+  const iFrameCommunicationPort = useRef<MessagePort>();
+
+  const hostMessageHandler = useCallback(
+    (event: MessageEvent<THostEventData>) => {
+      if (
+        event.data.eventName ===
+        CUSTOM_VIEWS_EVENTS_NAMES.CUSTOM_VIEW_INITIALIZATION
+      ) {
+        setHostContext(event.data.eventData.context);
+      } else {
+        console.warn(
+          `CustomViewShell: Unkown received event with name: ${event.data.eventName}`,
+          { event }
+        );
+      }
+    },
+    []
+  );
+
+  useEffect(() => {
+    const initializationMessageHandler = (event: MessageEvent) => {
+      console.log('Custom view shell received message', { event });
+
+      if (
+        event.origin === window.location.origin &&
+        event.data === CUSTOM_VIEWS_EVENTS_NAMES.CUSTOM_VIEW_BOOTSTRAP
```

Comment:
The only message we should listen in the global channel in the bootstrap one.
After that, we will listen for messages only in the private channel (the first one received in the `ports` array of the event)